Fixed bug where the 'Accept-Encoding' header could be improperly form…

danielkyu opened the pull request:
This is a PR that addresses the issue filed [here](https://github.com/adamfisk/LittleProxy/issues/342).

I'm not very familiar with the codebase but this is a first pass at how I would attempt to address this issue but I'm very much open to changes if need be.

Also, this is the first time I'm contributing to an open-source project so please let me know if I'm doing anything wrong.  Thanks :]

danielkyu committed:
Fixed bug where the 'Accept-Encoding' header could be improperly formatted after removing the 'sdch' encoding

danielkyu committed:
Fixed test failure issue

jekh commented on src/main/java/org/littleshoot/proxy/impl/ProxyUtils.java:
```diff
@@ -637,4 +637,30 @@ public static FullHttpResponse createFullHttpResponse(HttpVersion httpVersion,
+            // 'sdch' was the first element in the list and there are other
+            // encodings.
+            String noSdch = ae.replaceAll(",? *(sdch|SDCH)", "").replaceFirst("^ *, *", "");
+            headers.set(HttpHeaders.Names.ACCEPT_ENCODING, noSdch);
```
Would a subsequent call to `headers.getAll(HttpHeaders.Names.ACCEPT_ENCODING)` return a List of e.g. "gzip", "deflate", "br", or would it return a List with a single element of "gzip, deflate, br"?

The "preferred" way to manipulate headers with multi-value fields would be to use HttpHeaders.set, .add, etc. Unfortunately there is no HttpHeaders.remove(name, value), which is exactly what we want here. You could remove the sdch header from the List returned by getAll("sdch"), then call HttpHeaders.set("sdch", list).

danielkyu replied:
That's a great point; I didn't realize this issue.

The short answer to your question is that both scenarios are possible.  Here's an excerpt from my analysis:

```java
HttpHeaders headers = new DefaultHttpHeaders();

headers.add(HttpHeaders.Names.ACCEPT_ENCODING, "gzip");
headers.add(HttpHeaders.Names.ACCEPT_ENCODING, "deflate");
headers.add(HttpHeaders.Names.ACCEPT_ENCODING, "br");
headers.add(HttpHeaders.Names.ACCEPT_ENCODING, "foo, bar, baz");

// Prints "gzip" since it's the first value for the header.  Behavior follows from here: https://github.com/netty/netty/blob/netty-4.0.42.Final/codec-http/src/main/java/io/netty/handler/codec/http/DefaultHttpHeaders.java#L268-L289
System.out.println(headers.get(HttpHeaders.Names.ACCEPT_ENCODING));

// Prints "gzip", "deflate", "br" and "foo, bar, baz" each on its own line.  Behavior follows from here: https://github.com/netty/netty/blob/netty-4.0.42.Final/codec-http/src/main/java/io/netty/handler/codec/http/DefaultHttpHeaders.java#L297-L314
for (String encoding : headers.getAll(HttpHeaders.Names.ACCEPT_ENCODING)) {
  System.out.println(encoding);
}
```

Given that DefaultHttpHeaders.get() returns only the first header, I think this method should generally be avoided in favor of DefaultHttpHeaders.getAll() unless one knows for sure that there is, at most, one value mapped to the header.

For my specific case, I'm primarily concerned with how the headers are structured when I go through BrowserMob Proxy.  From what I was able to deduce, it appears that I'm being provided a single value to ACCEPT_ENCODING mapped to "gzip, deflate, sdch, br" as opposed to it being broken out individually for each encoding.  Unfortunately, that means I can't use the HttpHeaders.contains() method to check for the presence of the "sdch" encoding.

With all that said, if we want to properly support all the ways the encoding header can be defined either as...

1. Multiple distinct singular encodings (e.g. "gzip", "deflate", "br")
2. A single comma delineated string of encodings (e.g. "gzip, deflate, br")
3. Any combination of 1 and 2 (e.g. "gzip", "deflate, br")

...I propose the following:

1. Iterate through all the encodings that map to ACCEPT_ENCODING and construct a single comma delineated string of encodings.
2. Perform the current regex replacement that we have to remove any and all occurrences of "sdch".
3. Set the value of the ACCEPT_ENCODING header to the string that resulted after step 2.

What are your thoughts on this?

jekh replied:
Yep, you're absolutely right, my understanding of the behavior of HttpHeaders.getAll() was incorrect. I wish Netty provided just a bit more abstraction of the HttpHeaders object here.

I think your proposal is pretty solid. Perhaps you could use [Guava's Joiner](https://google.github.io/guava/releases/19.0/api/docs/com/google/common/base/Joiner.html) to join together all the headers returned by getAll(). Or you could loop through the Accept-Encoding headers one by one and run each one through the replaceAll() method.

👍

danielkyu replied:
Sweet.  I think your latter suggestion about looping through each of the encoders is the easiest/cleanest way to go.

jekh commented on src/main/java/org/littleshoot/proxy/impl/ProxyUtils.java:
```diff
@@ -637,4 +637,30 @@ public static FullHttpResponse createFullHttpResponse(HttpVersion httpVersion,
+            // encodings.
+            String noSdch = ae.replaceAll(",? *(sdch|SDCH)", "").replaceFirst("^ *, *", "");
+            headers.set(HttpHeaders.Names.ACCEPT_ENCODING, noSdch);
+            LOG.debug("Removed sdch and inserted: {}", noSdch);
```
I don't think there's any need to log this.

jekh commented on src/main/java/org/littleshoot/proxy/impl/ProxyUtils.java:
```diff
@@ -637,4 +637,30 @@ public static FullHttpResponse createFullHttpResponse(HttpVersion httpVersion,
+            LOG.debug("Removed sdch and inserted: {}", noSdch);
+        }
+
+        return headers;
```
This method should probably return null.

jekh reviewed (changes requested):
This looks pretty good, excellent first contribution, thank you. I have just a few review comments for you.

danielkyu committed:
Support for removing sdch encoding regardless of whether it is stored as a single/multi value

danielkyu commented:
@jekh Sorry for the delay.  I meant to get this out sooner but circumstances prevented me from doing so.

I should have addressed all your issues from the last PR and would appreciate a second go!

jekh commented:
Looks great, thanks @danielkyu. I'll go ahead and merge it.
